### DISCUSSION TITLE
Escaping `${{ }}` in workflows

### DISCUSSION THREAD

**Discussion by Saviq** [community/community#26621] [Actions]
<p>Hi, I’ve a need for literal <code>${{VAR}}</code> in a workflow. I’ve tried a number approaches to escaping the dollar sign, or the braces, but nothing seems to work:</p>
<pre><code class="lang-auto">${{{{VAR}}}}
\${{VAR}}
$\{{VAR}}
</code></pre>
<p>This either ends up with this error:</p>
<blockquote>
<p>The workflow is not valid. .github/workflows/var.yml (Line: 27, Col: 19): Unrecognized named-value: ‘VAR’ Located at position 1 within expression: VAR</p>
</blockquote>
<p>Or with the braces / backslashes remaining literal in the string.</p>
<p>I’ve not found any documentation about how to achieve this…</p>

**Comment by Simran-B** [comment, marked as answer]
<p>There might be tricks to avoid having the dollar sign next to the braces, but depending on the concrete situation it may or may not be possible to let it evaluate to <code>${{VAR}}</code> in the end.</p>
<p>The only generic solution I can think of is to use an expression that returns it as a string:</p>
<p><code>${{ '${{VAR}}' }}</code></p>

**Reply by jhm-ciberman** [reply]
That's a really creative solution!

**Reply by Pr0methean** [reply]
Does that mean there's no way to substitute an expression's value into another expression? 🤔

**Reply by Simran-B** [reply]
Nesting expressions is not supported but I haven't encountered a situation where it would be necessary to nest. Whenever I've seen users ask about nesting, it was solvable with format() and join() instead.

**Comment by Saviq** [comment]
<p>Thanks <a class="mention" href="/simran-b">@Simran-B</a>, that worked indeed!</p>

**Comment by 1oglop1** [comment]
Tried @Simran-B solution and got this `line 1: ${{VAR}}: bad substitution` so no luck

This worked for me
```
x=${{ '\${{VAR}}' }}
echo "$x"
```

**Reply by Simran-B** [reply]
That is peculiar, because it seems to be an error raised by the shell, but GitHub Actions expressions are evaluated before the shell code is run...

**Reply by divyegala** [reply]
@Simran-B I also get the `bad substitution` error with the previous answer, and with this comment I get `You have an error in your yaml syntax`. Any other ideas?

**Reply by Simran-B** [reply]
@1oglop1 @divyegala What do you actually want to achieve? To literally print `${{VAR}}` in Bash? What's the purpose, though?

The original question was how to prevent GitHub Actions from interpreting `${{ ... }}` as an expression, which can be achieved by using an expression that returns any string. If this is in a `run` block, the shell interprets whatever that string is, and `${{VAR}}` doesn't appear to be valid in Bash (I suppose the original question did not use Bash as the shell).

To print `${{VAR}}` with Bash as the shell, this does the trick: `${{ 'echo ''${{VAR}}''' }}`
Note that Bash requires **single** quote marks to not interpret `$` but treat it as literal text. Because the GitHub Actions expression uses single quote marks itself, you need to escape the `'` inside of the string by doubling it (`''`). What Bash receives is `echo '${{VAR}}'` in this case, which literally prints `${{VAR}}`.

**Reply by divyegala** [reply]
Thank you for the explanation @Simran-B. I actually wanted a variable expansion to occur in a bash script. I was able to use `${{ '${VAR}' }}` and that did the trick for me, by ensuring `VAR` was initialized in bash before the command that utilized it was run.

**Reply by tomhaynes** [reply]
We needed this for a "seed job" that generates other workflows.. I found I got `bad substitution` when trying the above, and needed to escape the internal `$` for it to work, i.e. -> `${{ '\${{ inputs.FOO }}' }}`

**Comment by leemeador** [comment]
I have successfully used this in the description of an action input in an action.yaml. The action runs.

Perhaps other informational strings in the action.yaml and/or generic workflows can use this:

```
description: 'Some info. Default: `\$\{\{ github.token \}\}`'
```

Note that:

- The string (after the colon) is surrounded by single quotes. 
- Inside that, the part in the dollar-double-curly-braces is enclosed in back tics. 
- And inside those, there are back-slashes escaping the dollar sign and all the curly braces.
